server/core/src/main/scala/org/infinispan/server/core/transport/ExtendedByteBuf.scala

```diff
@@ -72,4 +214,25 @@ object ExtendedByteBuf {
    def writeString(msg: Option[String], bf: ByteBuf) =
       writeRangedBytes(msg.map(_.getBytes(CharsetUtil.UTF_8)).getOrElse(Array.empty), bf)
 
+  sealed trait MaybeStringTrait {
+    def getValue(): String = throw new IllegalStateException("Value was not provided")
+  }
+
+  case object MoreBytesForString extends MaybeStringTrait
+  case object StringNotPresent extends MaybeStringTrait
+
+  final case class PresentString(value: String) extends MaybeStringTrait {
+    override def getValue(): String = value
+  }
+
+  sealed trait MaybeBytesTrait {
```

Comment:
This is a bit repetitive, you could take advantage of generics to define a single type (e.f. `Data[T]`) and then create `Data[Bytes]` and `Data[String]` versions, but we can always refactor this later on.